crates/egui/src/text_selection/text_cursor_state.rs

```diff
@@ -300,18 +300,97 @@ pub fn find_line_start(text: &str, current_index: CCursor) -> CCursor {
     // number of multi byte chars.
     // We need to know the char index to be able to correctly set the cursor
     // later.
-    let chars_count = text.chars().count();
+    let mut char_line_indexes: Vec<usize> = create_char_line_indexes(text);
+    let line_start_index = get_line_start_index(&mut char_line_indexes, current_index.index);
 
-    let position = text
-        .chars()
-        .rev()
-        .skip(chars_count - current_index.index)
-        .position(|x| x == '\n');
+    CCursor::new(line_start_index)
+}
+
+pub fn create_char_line_indexes(text: &str) -> Vec<usize> {
+    let mut line_indexes = vec![0];
 
-    match position {
-        Some(pos) => CCursor::new(current_index.index - pos),
-        None => CCursor::new(0),
+    let mut count = 0;
+    for (_i, c) in text.chars().enumerate() {
+        count += 1;
+        if c == '\n' {
+            line_indexes.push(count);
+        }
     }
+    line_indexes.push(count);
+
+    line_indexes
+}
+
+pub fn get_line_start_index(line_indexes: &mut [usize], index: usize) -> usize {
+    let (_current_line, line_start) = get_line_and_start_index(line_indexes, index);
+    line_start
+}
+
+pub fn get_line_and_start_index(line_indexes: &mut [usize], index: usize) -> (usize, usize) {
+    let mut current_line = 1;
+    for (i, line_start) in line_indexes.iter().enumerate() {
+        current_line = i;
+        if *line_start > index {
+            break;
+        }
+    }
+
+    (current_line, line_indexes[current_line - 1])
+}
+
+pub fn get_current_column(text: &str, line_indexes: &mut [usize], index: usize) -> usize {
+    let (current_column, _total_column) =
+        get_current_column_total_column(text, line_indexes, index);
+    current_column
+}
+
+pub fn get_current_column_total_column(
+    text: &str,
+    line_indexes: &mut [usize],
+    index: usize,
+) -> (usize, usize) {
+    let (current_column, total_column, _total_line) =
+        get_current_column_total_column_total_line(text, line_indexes, index);
+    (current_column, total_column)
+}
+
+pub fn get_current_column_total_column_total_line(
+    text: &str,
+    line_indexes: &mut [usize],
+    index: usize,
+) -> (usize, usize, usize) {
+    let line_start_index = get_line_start_index(line_indexes, index);
+
+    let total_lines = line_indexes.len() - 1;
+
+    let mut i: usize = line_start_index;
+    let mut current_column: usize = 1;
+    let mut total_columns: usize = 0;
+    while let Some(c) = text.chars().nth(i) {
```

Comment:
This has quadratic complexity, doesn't it? Each call to `nth()` is a for loop from the start of the string. You could do 

```rs
for c in text.chars().skip(line_start_index) {
```

Since you are already breaking on the newline. Otherwise you can get the index of the next line start and use `take()` to get the chars up to that index.

Reply:
This is to find a column according to the width of UTF8 characters.